blocks/autocompleters/index.js

```diff
@@ -0,0 +1,132 @@
+ */
+
+/**
+ * @callback FnOnSelect
+ * @param {*} value the value of the completer option.
+ * @param {Range} range the nodes included in the autocomplete trigger and query.
+ * @param {String} query the text value of the autocomplete query.
+ */
+
+/**
+ * @typedef {Object} Completer
+ * @property {?String} className A class to apply to the popup menu.
+ * @property {String} triggerPrefix the prefix that will display the menu.
+ * @property {FnGetOptions} getOptions get the block options in a resolved promise.
+ * @property {?FnAllowNode} allowNode filter the allowed text nodes in the autocomplete.
+ * @property {?FnAllowContext} allowContext filter the context under which the autocomplete activates.
+ * @property {FnOnSelect} onSelect
+ */
+
+/**
+ * Returns an "completer" definition for selecting from available blocks to replace the current one.
+ * The definition can be understood by the Autocomplete component.
+ *
+ * @param  {Function} onReplace  Callback to replace the current block.
+ * @returns {Completer}          Completer object used by the Autocomplete component.
+ */
+export function blockAutocompleter( { onReplace } ) {
+	const options = getBlockTypes().map( ( blockType ) => {
+		const { name, title, icon, keywords = [] } = blockType;
+		return {
+			value: name,
+			label: [
+				<BlockIcon key="icon" icon={ icon } />,
+				title,
+			],
+			keywords: [ ...keywords, title ],
+		};
+	} );
+
+	const getOptions = () => Promise.resolve( options );
+
+	const allowContext = ( before, range, after ) => {
+		return ! ( /\S/.test( before.toString() ) || /\S/.test( after.toString() ) );
+	};
+
+	const onSelect = ( blockName ) => {
+		onReplace( createBlock( blockName ) );
+	};
+
+	return {
+		className: 'blocks-block-autocomplete',
+		triggerPrefix: '/',
+		getOptions,
+		allowContext,
+		onSelect,
+	};
+}
+/**
+ * Returns a "completer" definition for inserting links to the posts of a user.
+ * The definition can be understood by the Autocomplete component.
+ *
+ * @returns {Completer} Completer object used by the Autocomplete component.
+ */
+export function userAutocompleter() {
+	const getOptions = () => {
+		return ( new wp.api.collections.Users() ).fetch().then( ( users ) => {
+			return users.map( ( user ) => {
+				return {
+					value: user,
+					label: [
+						<img key="avatar" alt="" src={ user.avatar_urls[ 24 ] } />,
+						<span key="name" className="name">{ user.name }</span>,
+						<span key="slug" className="slug">{ user.slug }</span>,
+					],
+					keywords: [ user.slug, user.name ],
+				};
+			} );
+		} );
+	};
+
+	const allowNode = ( textNode ) => {
+		return textNode.parentElement.closest( 'a' ) === null;
+	};
+
+	const onSelect = ( user, range ) => {
+		const mention = document.createElement( 'a' );
+		mention.href = user.link;
+		mention.textContent = '@' + user.name;
+		range.insertNode( mention );
+		range.setStartAfter( mention );
+		range.deleteContents();
```

Comment:
I wonder if the `userAutocompleter` should be aware of what to do with the selected value. This forces us to add a `range` property here where we could just execute a TinyMCE command if we were to have an instance of it. 

Same could be said above for the blocks autocompleter, assuming there's only one action to be done when selecting a block: replacing.

What if the `onSelect` is always provided by the calling side:

```js
<Autocomplete key="editable" completers={ [
	blockAutocompleter( block => onReplace( block ) ),
	userAutocompleter( user => this.editor.insertContent( linkFromUser( user ) ) ),
] }>
```

Reply:
I guess the difficulty here is that we need to remove what we already typed. Maybe we should do this consistently before calling the `onSelect`

Reply:
The advantage of Range is that it is an editor agnostic API and in the past I have been told I can't assume that components will always work with TinyMCE. I specifically did not want the autocomplete code to make any assumptions about what would be done with the range.
Note that TinyMCE doesn't really care about us changing the content out from under it and it would be breaking the abstraction that Editable is suppose to provide.

Reply:
Also one of the advantages of using range is that TinyMCE knows how to work with ranges.
For example if TinyMCE was available as `this.editor` then you could do:
```javascript
(user, range) => {
  this.editor.selection.setRng( range );
  this.editor.insertContent( linkFromUser( user ) );
}
```